subsys/net/lib/http/http_client.c

```diff
@@ -434,15 +414,32 @@ static int http_wait_data(int sock, struct http_request *req)
 			(void)http_parser_execute(
 				&req->internal.parser,
 				&req->internal.parser_settings,
-				req->internal.response.recv_buf + offset,
+				req->internal.response.recv_buf,
 				received);
 		}
 
 		total_received += received;
-		offset += received;
 
-		if (offset >= req->internal.response.recv_buf_len) {
-			offset = 0;
+		if (req->internal.response.cb) {
+			if (!req->internal.response.message_complete) {
```

Comment:
IMHO it is always easier to read the code if there is no `!` in it (less mental twisting needed). So would recommend
```
if (req->internal.response.message_complete) {
    DATA_FINAL case
} else {
    DATA_MORE case
}
```